internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixReadOnlySchemaRepository.java

```diff
@@ -372,23 +434,28 @@ public SchemaEntry getSupersetOrLatestValueSchema(String storeName) {
 
   @Override
   public Optional<SchemaEntry> getSupersetSchema(String storeName) {
-    schemaLock.readLock().lock();
+    getSchemaLock().readLock().lock();
     try {
       fetchStoreSchemaIfNotInCache(storeName);
-      SchemaData schemaData = schemaMap.get(getZkStoreName(storeName));
+      SchemaData schemaData = getSchemaMap().get(getZkStoreName(storeName));
       if (schemaData == null) {
         throw new VeniceNoStoreException(storeName);
       }
-      Optional<Integer> supersetSchemaID = getSupersetSchemaID(storeName);
-      return supersetSchemaID.map(schemaData::getValueSchema);
+      Optional<Integer> supersetSchemaIdOptional = getSupersetSchemaID(storeName);
+      supersetSchemaIdOptional.ifPresent(
+          supersetSchemaID -> maybeForceRefreshSchemaDataForSupersetSchemaWithRetry(
```

Comment:
This check seems bit heavy even for happy case particularly for derived and RMD schema find. can we cache the result ?

Reply:
Hmm I think it is actually fine, for the usage of this call, it will always call getLatestUpdateSchema which is also the same action, same for the A/A WC case. It is how it is working now and my understanding is it will just be a few check over memory objects...
I believe we can leave it for future optimization if we think this call is heavy - like we do see store with hundreds of schema, each containing more than one version of the update schema/rmd schema.

Reply:
So this function is in a hot path of applying write compute..

Maybe it is worth maintaining a flag in this class to cache the checked superset schema id per store.

Reply:
I did not maintain the exact ID here as I don't think it is going to be used anywhere in the check, I create a map of int->boolean to validate this. Please feel free to point out if that is okay.

Reply:
For today I think it is fine...but maybe in the future we want the latest version of it..

Reply:
I didn't see the code change in the latest PR.

I think having a map between store name and the validated superset schema id is good enough since we only care about the latest superset schema id.
If there is a new latest superset schema id, we could update that map after validation, right?

Reply:
Here we are talking about checking the existence of update schema / RMD schema of the superset schema as it iterates all derived schemas / RMD schemas.
The change is in SchemaData, where I add two maps from valueSchema ID to boolean. 
Since we don't use the latest derived schema id / RMD schema id there, it can essentially be two HashSet, given that SchemaData is not designed to be thread-safe that's also fine.

What do you think?

Reply:
For the checked superset schema id, then isn't potentially be the same as "stale" superset schema id?

Reply:
Yes for old schema id, but for new one, it will be fresh. The issue was new one was not found in the schema repo. Once we freshed the repo, we can keep using it by checking against the cache.

Reply:
Also forceRefresh should be guarded by writer lock I think

Reply:
if this is the case, then there is no need to keep the two set

Reply:
Right, no need for those sets.